src/optics.jl

```diff
@@ -318,6 +323,185 @@ function _modify(f, obj, r::Recursive, ::ModifyBased)
+    fns = fieldnames(O)
+    for (i, fn) in enumerate(fns)
+        v = Symbol("val$i")
+        st = if S <: ContextState 
+            if O <: Tuple 
+                :(ContextState(state.vals, obj, StaticInt{$(QuoteNode(fn))}()))
+            else
+                :(ContextState(state.vals, obj, StaticSymbol{$(QuoteNode(fn))}()))
+            end
+        else
+            :state
+        end
+        ms = :(($v, state) = f(getfield(props, $(QuoteNode(fn))), $st))
+        push!(modifications, ms)
+        push!(vals.args, v)
+    end
+    patch = O <: Tuple ? vals : :(NamedTuple{$fns}($vals))
+    start = :(props = getproperties(obj))
+    rest = MacroTools.@q begin
+        patch = $patch
+        new_obj = maybesetproperties(state, obj, patch)
+        return (new_obj, state)
+    end
+    Expr(:block, start, modifications..., rest) 
+end
+
+maybesetproperties(state, obj, patch) = setproperties(obj, patch)
+maybesetstate(state, obj, patch) = state
+
+abstract type AbstractQuery end
+
+"""
+    Query(select, descend, optic)
+    Query(; select=Any, descend=x -> true, optic=Properties())
+
+Query an object recursively, choosing fields where `select` 
+returns `true`, and descending when `descend` returns `true`.
+
+```jldoctest
+julia> using Accessors
+
+julia> q = Query(; select=x -> x isa Int, descend=x -> x isa Tuple)
+Query{var"#5#7", var"#6#8", Properties}(var"#5#7"(), var"#6#8"(), Properties())
+
+julia> obj = (7, (a=17.0, b=2.0f0), ("3", 4, 5.0), ((x=19, a=6.0,)), [1])
+(7, (a = 17.0, b = 2.0f0), ("3", 4, 5.0), (x = 19, a = 6.0), [1])
+
+julia> q(obj)
+(7, 4)
+```
+$EXPERIMENTAL
+"""
+struct Query{Select,Descend,Optic<:Union{ComposedOptic,Properties}} <: AbstractQuery
+    select_condition::Select
+    descent_condition::Descend
+    optic::Optic
+end
+Query(select, descend=x -> true) = Query(select, descend, Properties())
+Query(; select=Any, descend=x -> true, optic=Properties()) = Query(select, descend, optic)
+
+OpticStyle(::Type{<:AbstractQuery}) = SetBased()
+
+struct ContextState{V,O,FN}
+    vals::V
+    obj::O
+    fn::FN
+end
+struct GetAllState{V}
+    vals::V
+end
+struct SetAllState{C,V,I}
+    change::C
+    vals::V
+    itr::I
+end
+
+const GetStates = Union{GetAllState,ContextState}
+
+@inline pop(x) = first(x), Base.tail(x)
+@inline push(x, val) = (x..., val)
+@inline push(x::GetAllState, val) = GetAllState(push(x.vals, val))
+@inline push(x::ContextState, val) = ContextState(push(x.vals, val), nothing, nothing)
+
+(q::Query)(obj) = _getall(obj, q)
+
+function _getall(obj, q::Q) where Q<:Query
+    initial_state = GetAllState(())
+    _, final_state = let q=q
+        modify_stateful((obj, initial_state), q) do o, s
+            new_state = push(s, outer(q.optic, o, s))
+            o, new_state
+        end
+    end
+    final_state.vals
+end
+
+function set(obj, q::Q, vals) where Q<:Query
+    initial_state = SetAllState(Unchanged(), vals, 1)
+    final_obj, _ = let obj=obj, q=q, initial_state=initial_state
```

Comment:
These variables are not captured, but they are function arguments directly. So no need for let block?

Reply:
ah right, q is captured.